### ISSUE TITLE
Make Loader Optional 

### ISSUE THREAD

**Issue by fxck** [ngneat/transloco#38]
I didn't want to spam the other issue, but it's a continuation of [this](https://github.com/ngneat/transloco/issues/21#issuecomment-524838238).

https://stackblitz.com/edit/angular-tuqsp2?file=src/translations/translations.module.ts

What I'm trying to do:

- to have `TranslationModule`, to be imported in root module (AppModule), this module should 
  a. do transloco provider configuration 
  b. use `translocoService.setTranslation()` to set `common` translations, under `common` scope, doing:
```typescript
  this._translocoService.setTranslation({
    common: {
      common: 'This is common'
    }
  }, 'en', { merge: true });
```
       c. export `TranslocoModule` for transloco components used inside AppComponent template
- to have <Feature>TranslationModule, to be imported inside the lazy loaded `<Feature>Module`, this should
  a. use `translocoService.setTranslation()` to merge `<feature>` translation object, doing
```typescript
  this._translocoService.setTranslation({
    <feature>: {
      lazy: 'This is lazy loaded'
    }
  }, 'en', { merge: true });
```
       c. export `TranslocoModule` for transloco components used inside <Feature>Component template

Now first this that happens is that I do not want to use http to load translation files at all, but if I dont't provide `TRANSLOCO_LOADER`, I get 

> NullInjectorError: No provider for InjectionToken TRANSLOCO_LOADER!

so to satisfy it, I use
```typescript
@Injectable({ providedIn: 'root' })
export class TempTranslationLoader implements TranslocoLoader {
  getTranslation(langPath: string) {
    return of({});
  }
}
```

what happens is this

![image](https://user-images.githubusercontent.com/1303561/63704005-63335180-c82a-11e9-9c99-0b724edc55ec.png)

```
🍻 Translation Load Success: en
Missing translation for 'common.ok' 🤔🕵🏻‍♀
Missing translation for 'foo' 🤔🕵🏻‍♀
Missing translation for 'common.confirm' 🤔🕵🏻‍♀
Missing translation for 'foo' 🤔🕵🏻‍♀
```

so the common translations (neither nested, nor in root) do not work, but the lazy loaded does

I tried putting the `setTranslation` on various places, didn't work, neither did using setTimeouts..

@NetanelBasal @shaharkazaz

**Comment by itayod** [comment]
Hi, @fxck I believe the reason it didn't work for you is that you use `setTranslation` before the language is loaded:

I added this to the constructor and it works well:

```typescript
this._translocoService.events$.subscribe((e) => {
      if(e.type === 'translationLoadSuccess' && e.payload.lang === 'en') {
        this._translocoService.setTranslation({
          common: {
            ok: 'Ok',
            confirm: 'Confirm'
          }
        }, 'en', { merge: true });
        
      }
    })
```

**Comment by fxck** [comment]
Guess it boils down to making the loader optional then? Language could be loaded / registered by the first `setTranslation` with given language, maybe?

**Comment by shaharkazaz** [comment]
@fxck I believe you are correct, making the loader is something we can think about @NetanelBasal what do you think?
But I think for now you can implement a basic loader and use scopes as you need when #33 is merged.
Check out this #24 discussion and see if the PR will you with your issue.

**Comment by fxck** [comment]
> Check out this #24 discussion and see if the PR will you with your issue.

I think of the way I described as of an alternative to letting the library asynchronously load translation files and do the scoping for me. So even after #33 is merged I'd rather do it this way..

The only main con (for my use case anyway) is that I'm bundling all languages with the module, instead of only loading the one needed, but the pro is not having to deal with another loading state,  having those translations inside the feature folder, rather than having them in assets and to be able to easily use backticks and comments, given that we do not support many languages and that the app is split quite vigorously into lazy loaded modules I don't think the file size will be much of a problem. 

Also I think you get to do this https://dev.to/andreasbergqvist/typescript-get-types-from-data-using-typeof-4b9c and thus getting types for free (haven't yet tried, but it should work in theory)

**Comment by shaharkazaz** [comment]
@fxck Sounds legit to me. No worries, we'll add it this week.

**Comment by shaharkazaz** [comment]
@fxck I was wondering, how do you handle language changes in your project? 
Based on your example the language seems static.

**Comment by fxck** [comment]
It's an app, url stays the same, so on the fly as soon as users saves his settings, without reloading.

**Comment by fxck** [comment]
That setTranslation in translation modules is repeated for each language.

**Comment by fxck** [comment]
What I'm thinking about doing is something like this

```typescript

// base language
const en = { title: 'Title' };

// get types so no keys would be lost in other translations
export type Types = typeof en;

const cs: Types = { title: 'Titulek' };
const pl: Types = { title: 'Polish title' };


export class SomeTranslationModule {
  constructor (private _myTranslationService: MyTranslationService) { 
    this._myTranslationService.setTranslations({ en, cs, pl });
  }
}

```

**Comment by fxck** [comment]
I guess `Types` could also be used for typesafe translation keys directly in the template with something like..

```typescript
// if selectTranslate supported generic type
translations$ = this.service.selectTranslate<Types>('uniqueKey');
```

```html
<ng-container *ngFor="let t of [translations$ | async]">
  {{ t?.title }}
</ng-container>
```

..instead of using the `transloco` structural directive (that [`*ngFor`](https://github.com/angular/angular/issues/15280#issuecomment-508012536) explained here). Not the most convenient but yeah.

**Comment by shaharkazaz** [comment]
@fxck So if I understand correctly, you are preloading every language to the bundle? did I understand correctly?

**Comment by fxck** [comment]
Yep. My reasoning behind it is described here https://github.com/ngneat/transloco/issues/38#issuecomment-524983909

**Comment by NetanelBasal** [comment]
Available in 1.5.0
